### PR TITLE
Clipboard data API revamp

### PR DESCRIPTION
**Pull request by slouken**
The clipboard data API is now supported on all platforms, at least for internal use.

### CODE REVIEW

**Comment by slouken** [comment]
@LiquidityC, can you take a look? I revamped the API a bit and made it work, at least internally, for all platforms.

**Comment by slouken** [comment]
One of the changes I made was, in addition to the event, the callback is called with NULL when the clipboard is cleared, so the same function that handles requests can also handle cleanup, without having to process events. This is especially useful when shutting down, as event processing could potentially be stopped.

Another idea is to pass a cleanup function (or NULL) to SDL_SetClipboardData() to be more explicit in the cleanup process. I think I like that better. Thoughts?

File: include/SDL3/SDL_clipboard.h [hunk 1]
```diff
@@ -147,7 +150,7 @@ extern DECLSPEC SDL_bool SDLCALL SDL_HasPrimarySelectionText(void);
  *
  * \sa SDL_SetClipboardData
  */
-typedef void *(SDLCALL *SDL_ClipboardDataCallback)(size_t *size, const char *mime_type, void *userdata);
+typedef void *(SDLCALL *SDL_ClipboardDataCallback)(void *userdata, const char *mime_type, size_t *size);
```
Comment (sezero) [inline include/SDL3/SDL_clipboard.h:153]: Was this truly necessary? We'll have to intercept this in sdl2-compat. (same for `SDL_SetClipboardData`)

Comment (slouken) [reply include/SDL3/SDL_clipboard.h:153]: This API isn't in SDL2, it's new for SDL 3.0.

**Comment by LiquidityC** [comment]
I can take a look this weekend.

**Comment by slouken** [comment]
> Another idea is to pass a cleanup function (or NULL) to SDL_SetClipboardData() to be more explicit in the cleanup process. I think I like that better. Thoughts?

I ended up doing this, and also adding a Windows implementation.

File: src/stdlib/SDL_string.c [hunk 2]
```diff
@@ -1774,6 +1774,20 @@ static size_t SDL_PrintFloat(char *text, size_t maxlen, SDL_FormatInfo *info, do
...
+static size_t SDL_PrintPointer(char *text, size_t maxlen, SDL_FormatInfo *info, const void *value)
+{
+    char num[130];
+    size_t length;
+
+    if (!value) {
+        return SDL_PrintString(text, maxlen, info, NULL);
+    }
+
+    SDL_ulltoa((unsigned long long)(uintptr_t)value, num, 16);
```
Comment (sezero) [inline src/stdlib/SDL_string.c:1786]: I wonder whether this works correctly on 128 bit pointers (i.e. Arm's CHERI: https://github.com/libsdl-org/SDL/pull/4552)

Also this particular patch needed in SDL2 too ?

Comment (slouken) [reply src/stdlib/SDL_string.c:1786]: I believe unsigned long long will be 128 bits on that architecture.

Yes, this is a good patch for SDL2.

Comment (slouken) [reply src/stdlib/SDL_string.c:1786]: Done!

**Review by LiquidityC — COMMENTED** [review]
Looks good

**Comment by slouken** [comment]
Thanks!

**Pull request merged**